model/task/dependency_graph.go

```diff
@@ -0,0 +1,272 @@
+package task
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/multi"
+	"gonum.org/v1/gonum/graph/topo"
+	"gonum.org/v1/gonum/graph/traverse"
+)
+
+// DependencyGraph models task dependency relationships as a directed graph.
+// Use NewDependencyGraph to initialize a new DependencyGraph.
+type DependencyGraph struct {
+	transposed          bool
+	graph               *multi.DirectedGraph
+	tasksToNodes        map[TaskNode]graph.Node
+	nodesToTasks        map[graph.Node]TaskNode
+	edgesToDependencies map[edgeKey]DependencyEdge
+}
+
+// NewDependencyGraph returns an initialized DependencyGraph.
+// transposed determines the direction of edges in the graph.
+// If transposed is false, edges point from dependent tasks to the tasks they depend on.
+// If transposed is true, edges point from depended on tasks to the tasks that depend on them.
+func NewDependencyGraph(transposed bool) DependencyGraph {
+	return DependencyGraph{
+		transposed:          transposed,
+		graph:               multi.NewDirectedGraph(),
+		tasksToNodes:        make(map[TaskNode]graph.Node),
+		nodesToTasks:        make(map[graph.Node]TaskNode),
+		edgesToDependencies: make(map[edgeKey]DependencyEdge),
+	}
+}
+
+type edgeKey struct {
+	from TaskNode
+	to   TaskNode
+}
+
+// DependencyEdge is a representation of a dependency in the graph.
+type DependencyEdge struct {
+	// Status is the status specified by the dependency, if any.
+	Status string
+	// From is the node the edge begins from.
+	From TaskNode
+	// To is the node the edge points to.
+	To TaskNode
+}
+
+// TaskNode is the representation of a task in the graph.
+type TaskNode struct {
+	// Name is the display name of the task.
+	Name string
+	// Variant is the build variant of the task.
+	Variant string
+	// ID is the task's ID.
+	ID string
+}
+
+// String represents TaskNode as a string.
+func (t TaskNode) String() string {
+	if t.ID != "" {
+		return t.ID
+	}
+
+	return fmt.Sprintf("%s/%s", t.Variant, t.Name)
+}
+
+// VersionDependencyGraph finds all the tasks from the version given by versionID and constructs a DependencyGraph from them.
+func VersionDependencyGraph(versionID string, transposed bool) (DependencyGraph, error) {
```

Comment:
Do we anticipate using this? (Will we remember to remove this if we don't?)

Reply:
removed. We'll use it eventually, but I can put it back then.
Though `buildFromTasks` is extra now it's used by a bunch of tests and rewriting them would be a lot.